File: solutions/238/238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        int n = nums.size();

        vector<int> prefix(n, 1);
        prefix[0] = 1;
        for(int i = 1; i < nums.size(); i++) {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        vector<int> sufix(n, 1);
        sufix[nums.size() - 1] = 1;
        for(int i = nums.size() - 2; i >= 0; i--) {
            sufix[i] = sufix[i + 1] * nums[i + 1];
        }

        vector<int> ans(n, 1);
        for(int i = 0; i < ans.size(); i++) {
            ans[i] = prefix[i] * sufix[i];
        }

        return ans;
    }
};

File: solutions/238/238.swift
class Solution {
    func productExceptSelf(_ nums: [Int]) -> [Int] {
        let n = nums.count
        
        var prefix = Array(repeating: 1, count: n)
        prefix[0] = 1
        for i in 1..<nums.count {
            prefix[i] = prefix[i - 1] * nums[i - 1]
        }
        
        var sufix = Array(repeating: 1, count: n)
        sufix[nums.count - 1] = 1
        for i in stride(from: nums.count - 2, through: 0, by: -1) {
            sufix[i] = sufix[i + 1] * nums[i + 1]
        }
        
        var ans = Array(repeating: 1, count: n)
        for i in 0..<ans.count {
            ans[i] = prefix[i] * sufix[i]
        }
        
        return ans
    }
}